fix(qe): fix error logging

miguelff opened the pull request:
Fixes #16226 

This patch ensures that errors resulting in a failed query generate an event if the error logger is enabled.  

This aligns with the principle of least astonishment: The user expects errors to be logged, and due to [the lack of a formal definition of what constitutes an error](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#log-levels), the most sensible approach in my opinion (backed by bug reports like https://github.com/prisma/prisma/issues/14933) is to log any condition leading to an error in the processing of the query, which is, in turn, any of [the four well-known classes of errors](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#log-levels).

This is done now [in the RequestHandler](https://github.com/prisma/prisma/pull/16287/files#diff-41021953e5dbe1050efb578ed85e14befe861bce256f4c312fac4bef801e3731R188-R203) thus capturing any possible error condition happening during query processing. Before, every different engine was doing this differently, and errors were not properly captured.

A smoke test suite proves that `NotFoundError` [errors in `findXOrThrow` are now properly emitted](https://github.com/prisma/prisma/issues/14933). I consider this enough, given that the code that handles the error is the same for every exception thrown during query execution.

miguelff committed:
Remove confusing comment

miguelff committed:
Add support for single point of error logging

In case error logging is enabled in the client, a middleware
captures errors and uses the event emitter to notify about them. This way, errors are captured in the same way accross the different engines, by capturing errors one level above in the call stack (at the client level)

After this approach is validated, next commits will:
- Filter the kind of errors that should be logged
- Create the appropriate structure for the logging event being emitted
- Add unit tests to check the behavior with the binary engine.

miguelff committed:
log at the request level instead of through a middleware

Our middleware is FIFO (a queue rather than an actual stack) so the first middleware would call on other middlewares defined by the user before hitting the actual request. As we don't want to observe on users's middleware errors, I moved the logic from a middleware to the point where the request is being made.

Also notice we don't want to block on user handling code upon request resolving, so the error handling code runs asynchronously. This has a side effect. If the execution of the program halts, before the promise is resolved in the next event loop tick, then the error won't be logged. This might not be an acceptable trade-off, and I might change it in a follow-up commit.

miguelff committed:
Blocking on handling code as it doesn't change elpased time of queries

miguelff committed:
Ignore unhandled errors with a default handler

miguelff committed:
Update RequestHandler.ts

miguelff commented on packages/client/src/runtime/getPrismaClient.ts:
```diff
@@ -339,12 +340,17 @@ export function getPrismaClient(config: GetPrismaClientConfig) {
         validatePrismaClientOptions(optionsArg, config.datasourceNames)
       }
 
+      this._logEmitter = new EventEmitter()
```
Different engines have their own instance of the EventEmitter, that for some reason is not a singleton. I thought it made more sense to inject the dependency, as the event pub/sub should be the same across the client codebase.

Then, wherever else it is used, it's injected.

miguelff committed:
Don't export node type in PrismaClient

miguelff committed:
Do not depend on Node type

miguelff committed:
Omit error, and instead warn when retrying

Error will flow up the stack and be captured by the RequestHandler

miguelff committed:
Remove unnecessary event emission from  libraryEngine

miguelff committed:
Refactor binary engine error logic for clarity

miguelff committed:
Emit a clearer event

miguelff committed:
Qualify binary engine error target

miguelff commented on packages/engine-core/src/common/errors/PrismaClientRustError.ts:
```diff
@@ -1,28 +1,27 @@
 import type { PrismaClientRustErrorArgs } from './types/PrismaClientRustErrorArgs'
-import { getBacktraceFromLog, getBacktraceFromRustError } from './utils/log'
```
Changes in this file, need to be reviewed together with changes in [log.ts](https://github.com/prisma/prisma/pull/16287/files#diff-7cdf39e07157969396116c74b51ad1ab01705f3d2691f9e1a6aa60ba1007846e) and [some changes in BinaryEngine.ts] which were aimed at simplifying error processing logic during engine initialization, which remains functionally untouched, but a bit easier to understand. 

To refactor that part of BinaryEngine, I only made the `PrismaClientRustError` treat an error uniformly from an ErrorLog trace than from an actual Panic, and for that, I hid specific methods that dealt with `RustLog` and `RustError` and make equivalent methods that dealt with the union type exported.

miguelff committed:
Tests

danstarns commented on packages/client/src/runtime/getPrismaClient.ts:
```diff
@@ -342,6 +343,10 @@ export function getPrismaClient(config: GetPrismaClientConfig) {
       }
 
+      const logEmitter = new EventEmitter().on('error', () => {
+        // to prevent unhandled error events
```
Should we at least log or DEBUG these?

miguelff replied:
Hey @danstarns I was about to a debug trace in here like suggested, but then ended up writing a clarification comment and not doing it. Let me see if it makes sense.

From https://github.com/prisma/prisma/pull/16287/commits/a13c40cf1d4a44c2bc11b9b5039cd4a678b9b432

```typescript
      const logEmitter = new EventEmitter().on('error', (e) => {
        // this is a no-op to prevent unhandled error events
        //
        // If the user enabled error logging this would never be executed. If the user did not
        // enabled error logging, this would be executed, and a trace for the error would be logged
        // in debug mode, which is like going in the opposite direction than what the user wanted by
        // not enabling error logging in the first place.
      })
```

danstarns commented on packages/client/tests/functional/query-error-logging/_matrix.ts:
```diff
@@ -0,0 +1,24 @@
+import { defineMatrix } from '../_utils/defineMatrix'
+
+export default defineMatrix(() => [
+  [
+    {
+      provider: 'sqlite',
+    },
+    {
+      provider: 'postgresql',
+    },
+    {
+      provider: 'mysql',
+    },
+    {
+      provider: 'mongodb',
+    },
+    {
+      provider: 'cockroachdb',
+    },
+    {
+      provider: 'sqlserver',
+    },
+  ],
+])
```
```suggestion
import { defineMatrix } from '../_utils/defineMatrix'
import { allProviders } from '../_utils/providers'

export default defineMatrix(() => [allProviders])
```

Commit:
Update packages/client/tests/functional/query-error-logging/_matrix.ts

Co-authored-by: Daniel Starns <danielstarns@hotmail.com>

miguelff commented:
Taking a look at the windows failures. Please don't let this hold you from a review.

miguelff committed:
Not import runtime in tests

miguelff committed:
Clarify comment on default error handler

millsp commented on packages/client/src/runtime/getPrismaClient.ts:
```diff
@@ -342,6 +343,15 @@ export function getPrismaClient(config: GetPrismaClientConfig) {
         validatePrismaClientOptions(optionsArg, config.datasourceNames)
       }
 
+      const logEmitter = new EventEmitter().on('error', (e) => {
```
In `eventemitter3`

> We do not throw an error when you emit an error event and nobody is listening.

miguelff replied:
Agreed that's one of the reasons why we should use eventemitter3 altogether and stop depending on Node's implementation.

millsp commented on packages/engine-core/src/binary/BinaryEngine.ts:
```diff
@@ -236,36 +234,40 @@ You may have to run ${chalk.greenBright('prisma generate')} for your changes to
+    this.logEmitter.emit('error', {
+      message: message,
+      timestamp: new Date(),
+      target: 'binary engine lifecycle management',
```
Is this intentional?

miguelff replied:
I made some changes based on https://github.com/prisma/prisma/pull/16287#discussion_r1029925661. This is done differently now.

millsp commented on packages/engine-core/src/data-proxy/DataProxyEngine.ts:
```diff
@@ -403,7 +397,9 @@ export class DataProxyEngine extends Engine {
         }
 
-        this.logEmitter.emit('warn', { message: 'This request can be retried' })
+        this.logEmitter.emit('warn', {
```
I don't think sentence was documented but let's check if that was the case

miguelff replied:
Good call: [It was not documented](https://www.google.com/search?q=%22can+be+retried%22+site%3Aprisma.io&ei=3tx9Y8T4DrWnkdUPyq2A8AM&ved=0ahUKEwjE37ib9MP7AhW1U6QEHcoWAD4Q4dUDCA8&uact=5&oq=%22can+be+retried%22+site%3Aprisma.io&gs_lcp=Cgxnd3Mtd2l6LXNlcnAQA0oECEEYAUoECEYYAFCvCFj0OWCXPGgCcAB4AoAB2AGIAY0MkgEGMTIuMy4xmAEAoAEBwAEB&sclient=gws-wiz-serp)

millsp commented on packages/engine-core/src/binary/BinaryEngine.ts:
```diff
@@ -236,36 +234,40 @@ You may have to run ${chalk.greenBright('prisma generate')} for your changes to
     }
   }
 
+  private emitError(err: any) {
```
(Warning: Potentially stupid question)
Why are we emitting errors here still vs always emitting in `RequestHandler`, is it for a specific set of errors that only happen in BinaryEngine? https://github.com/prisma/prisma/pull/16287/files#r1029921786

miguelff replied:
I have simplified the binary engine error workflow extensively by relying on the type system, in particular removing the union type `RustError`, consolidating the states: `lastPanic`, `lastError`, `lastErrorlog` into `lastError: PrismaClientRustError` which allowed me to remove a lot of conditionals: and finally getting rid of the error emission, in `setError` (now it just sets the error, handling a possible panic), and making it explicit in the places where the state of the server is managed. 

https://github.com/prisma/prisma/pull/16287/commits/00a497ed5ad82deebbcdcb09dd633bfb0c00d7ce

I will still need to do some manual testing on different error conditions. I would love to be able to automate things like I with https://github.com/Shopify/toxiproxy to introduce latency or network errors as part of the test suite, but I think that goes way beyond the scope of that issue.

miguelff committed:
Simplify BinaryEngine

This change is about:
1. getting rid of lastErrorLog, lastRustError, lastPanic
state and instead build all the error information when we parse a log line if that's pertinent

2. Let request errors flow up the stack instead to be emitted in the RequestHandler.

3. Only emit events explicitly during lifecycle management of the binary engine

miguelff commented:
Did the following:
- Tampered the binary engine:

```diff
diff --git a/query-engine/request-handlers/src/graphql/handler.rs b/query-engine/request-handlers/src/graphql/handler.rs
index 39cb2e08e1..15690c5b47 100644
--- a/query-engine/request-handlers/src/graphql/handler.rs
+++ b/query-engine/request-handlers/src/graphql/handler.rs
@@ -26,7 +26,6 @@ impl<'a> GraphQlHandler<'a> {
 
     pub async fn handle(&self, body: GraphQlBody, tx_id: Option<TxId>, trace_id: Option<String>) -> PrismaResponse {
         tracing::debug!("Incoming GraphQL query: {:?}", &body);
-
         match body.into_doc() {
             Ok(QueryDocument::Single(query)) => self.handle_single(query, tx_id, trace_id).await,
             Ok(QueryDocument::Multi(batch)) => match batch.compact(self.query_schema) {
@@ -44,7 +43,7 @@ impl<'a> GraphQlHandler<'a> {
 
     async fn handle_single(&self, query: Operation, tx_id: Option<TxId>, trace_id: Option<String>) -> PrismaResponse {
         use user_facing_errors::Error;
-
+        dbg!("Running single {}", query.clone());
         let gql_response = match AssertUnwindSafe(self.handle_graphql(query, tx_id, trace_id))
             .catch_unwind()
             .await
@@ -201,6 +200,9 @@ impl<'a> GraphQlHandler<'a> {
         tx_id: Option<TxId>,
         trace_id: Option<String>,
     ) -> query_core::Result<ResponseData> {
+        if 2 > 1 {
+            panic!("Forced")
+        }
         self.executor
             .execute(tx_id, query_doc, self.query_schema.clone(), trace_id)
             .await

```

- Ran Prisma against the binary engine. The connection is reset before the actual panic is processed asynchronously. In this case, the error is correctly received by the client

- Ran Prisma against the library engine. The behavior is correct.

I think this is ready to be merged.

aqrln commented on packages/engine-core/src/data-proxy/DataProxyEngine.ts:
```diff
@@ -382,10 +380,6 @@ export class DataProxyEngine extends Engine {
       try {
         return await args.callback({ logHttpCall })
       } catch (e) {
-        this.logEmitter.emit('error', {
```
I think we would lose the logging for non-retryable errors which are not request errors if we remove it (e.g. Data Proxy HTTP 400 errors), wouldn't we? Or is that not a problem since they'll be thrown anyway?

miguelff replied:
They will be thrown anyway, yes. Shouldn't be a problem.

aqrln reviewed (commented):
One comment/question, but other than that, looks very good to me! Thank you for the fix and for the nice refactoring!
